packages/fx-core/src/core/middleware/projectUpgrader.ts

```diff
@@ -0,0 +1,234 @@
+  await upgradeContext(ctx);
+  await next();
+};
+
+// This part is for update context and userdata file to support better local debug experience.
+export async function upgradeContext(ctx: CoreHookContext): Promise<void> {
+  try {
+    const inputs = ctx.arguments[ctx.arguments.length - 1] as Inputs;
+    if (!inputs.projectPath) {
+      ctx.result = err(NoProjectOpenedError());
+      return;
+    }
+    const projectPathExist = await fs.pathExists(inputs.projectPath);
+    if (!projectPathExist) {
+      ctx.result = err(PathNotExistError(inputs.projectPath));
+      return;
+    }
+    const confFolderPath = path.resolve(inputs.projectPath!, `.${ConfigFolderName}`);
+    const settingsFile = path.resolve(confFolderPath, "settings.json");
+    const projectSettings: ProjectSettings = await fs.readJson(settingsFile);
+    if (!projectSettings.currentEnv) {
+      projectSettings.currentEnv = "default";
+    }
+
+    // Read context file.
+    const contextPath = path.resolve(confFolderPath, `env.${projectSettings.currentEnv}.json`);
+    const context = await readContext(contextPath);
+
+    // Update value of specific key in context file to secret pattern.
+    // Return: map of updated values.
+    const updatedKeys = updateContextValue(context);
+    if (!updatedKeys || updatedKeys.size == 0) {
+      // No keys need to be updated, which means the file is up-to-date.
+      // Can quit directly.
+      return;
+    }
+
+    // Some keys updated.
+    // Save the updated context and send log.
+    await saveContext(contextPath, context);
+    const core = ctx.self as FxCore;
+    core?.tools?.logProvider?.info(
+      "[core]: template version is too low. Will update context and move some configs from env to userdata."
+    );
+
+    // Read UserData file.
+    const userDataPath = path.resolve(confFolderPath, `${projectSettings.currentEnv}.userdata`);
+    const userData = await readUserData(userDataPath, projectSettings.projectId);
+
+    // Merge updatedKeys into UserData.
+    mergeKeysToUserDate(userData, updatedKeys);
+
+    // Save UserData
+    await saveUserData(userDataPath, userData, projectSettings.projectId);
+  } catch (error) {
+    ctx.result = err(ContextUpgradeError(error));
+  }
+}
+
+async function readUserData(
+  userDataPath: string,
+  projectId?: string
+): Promise<Record<string, string>> {
+  let dict: Record<string, string> = {};
+  if (await fs.pathExists(userDataPath)) {
+    const dictContent = await fs.readFile(userDataPath, "UTF-8");
+    dict = deserializeDict(dictContent);
+  }
+
+  if (projectId) {
+    const cryptoProvider = new LocalCrypto(projectId);
+    for (const secretKey of Object.keys(dict)) {
+      if (!dataNeedEncryption(secretKey)) {
+        continue;
+      }
+      const secretValue = dict[secretKey];
+      const plaintext = cryptoProvider.decrypt(secretValue);
+      if (plaintext.isErr()) {
+        throw plaintext.error;
```

Comment:
decrypt one field error will cause the whole load fail?

Reply:
Following logic from @tecton.
If decrypt failed, userdata can not be load correctly, thus I also think we should fail the process here.